staging/src/github.com/openshift/openshift-controller-manager/pkg/build/controller/build/build_controller.go

```diff
@@ -360,10 +373,55 @@ func (bc *BuildController) defaults() builddefaults.BuildDefaults {
 	return copy
 }
 
-func (bc *BuildController) setBuildDefaultProxy(spec *configv1.ProxySpec) {
+func (bc *BuildController) setBuildProxy(buildCfg *configv1.Build, globalProxy *configv1.Proxy) {
 	bc.configLock.Lock()
 	defer bc.configLock.Unlock()
-	bc.buildDefaults.DefaultProxy = spec
+
+	if buildCfg == nil && globalProxy == nil {
+		return
+	}
+
+	cfg := bc.buildDefaults.Config
+	if cfg == nil {
+		cfg = &openshiftcontrolplanev1.BuildDefaultsConfig{}
+		bc.buildDefaults.Config = cfg
+	}
+	defaultProxy := bc.buildDefaults.DefaultProxy
+	if defaultProxy == nil {
+		defaultProxy = &configv1.ProxySpec{}
+		bc.buildDefaults.DefaultProxy = defaultProxy
+	}
+
+	// first apply the global proxy settings
+	if globalProxy != nil {
+		bc.setBuildDefaultProxy(defaultProxy, globalProxy.Status.HTTPProxy, globalProxy.Status.HTTPSProxy, globalProxy.Status.NoProxy)
+		bc.setBuildGitProxy(cfg, globalProxy.Status.HTTPProxy, globalProxy.Status.HTTPSProxy, globalProxy.Status.NoProxy)
+	}
+
+	// then build config takes precedence
+	if buildCfg != nil {
+		// if default set, set both default/build and git proxy
+		if buildCfg.Spec.BuildDefaults.DefaultProxy != nil {
+			bc.setBuildDefaultProxy(defaultProxy, buildCfg.Spec.BuildDefaults.DefaultProxy.HTTPProxy, buildCfg.Spec.BuildDefaults.DefaultProxy.HTTPSProxy, buildCfg.Spec.BuildDefaults.DefaultProxy.NoProxy)
+			bc.setBuildGitProxy(cfg, buildCfg.Spec.BuildDefaults.DefaultProxy.HTTPProxy, buildCfg.Spec.BuildDefaults.DefaultProxy.HTTPSProxy, buildCfg.Spec.BuildDefaults.DefaultProxy.NoProxy)
+		}
+		// but if git proxy specifically set, override any settings of git proxy from default proxy
+		if buildCfg.Spec.BuildDefaults.GitProxy != nil {
+			bc.setBuildGitProxy(cfg, buildCfg.Spec.BuildDefaults.GitProxy.HTTPProxy, buildCfg.Spec.BuildDefaults.GitProxy.HTTPSProxy, buildCfg.Spec.BuildDefaults.GitProxy.NoProxy)
+		}
+	}
+}
+
+func (bc *BuildController) setBuildDefaultProxy(defaultProxy *configv1.ProxySpec, httpProxy, httpsProxy, noProxy string) {
+	defaultProxy.HTTPProxy = httpProxy
```

Comment:
We could probably simplify things a bit if we just update the config values on the build controller:
```
bc.buildDefaults.DefaultProxy.HTTPProxy = httpProxy
...
```

Reply:
addressed in next push